### PR TITLE
plan, executor: improve merge join to support all join types

### LINKED ISSUE #4822: improve merge join to support all join types
Currently, merge join only support `LeftOuterJoin`, `RightOuterJoin`, `InnerJoin`, the following join types are not supported:
- `SemiJoin`
- `AntiSemiJoin`
- `LeftOuterSemiJoin`
- `AntiLeftOuterSemiJoin`

### PR DESCRIPTION
**Pull request by zz-jason**
fix #4822 

main changes:
1. Extracted a `joinResultGenerator` to file **executor/join_result_generators.go** to emit join results according to the concrete join type (SemiJoin, AntiSemiJoin, LeftOuterJoin, InnerJoin ...), some branch predicating is removed
2. Extracted a `joinBuilder` to file **executor/join_builder.go**
3. Implemented **SemiJoin**, **Anti SemiJoin**, **LeftOuterSemiJoin**, **Anti LeftOuterSemiJoin** in `MergeJoinExec`
4. Rewrited `buildSchema()` to handle `SemiJoin` and `LeftOuterSemiJoin` in `PhysicalMergeJoin`

### CODE REVIEW

File: plan/explain_test.go [hunk 1]
```diff
@@ -237,12 +237,12 @@ func (s *testExplainSuite) TestExplain(c *C) {
...
-				"TableReader_19 HashSemiJoin_16  root data:TableScan_18 8000",
-				"TableScan_14   cop table:t2, range:(-inf,+inf), keep order:false 8000",
-				"TableReader_15 HashSemiJoin_16  root data:TableScan_14 8000",
-				"HashSemiJoin_16 StreamAgg_9 TableReader_19,TableReader_15 root right:TableReader_15, aux, equal:[eq(test.t1.c1, test.t2.c1)] 8000",
-				"StreamAgg_9  HashSemiJoin_16 root type:stream, funcs:sum(5_aux_0) 1",
+				"TableScan_13   cop table:t1, range:(-inf,+inf), keep order:true 8000",
+				"TableReader_14 MergeJoin_27  root data:TableScan_13 8000",
+				"IndexScan_21   cop table:t2, index:c1, range:[<nil>,+inf], out of order:false 8000",
+				"IndexReader_22 MergeJoin_27  root index:IndexScan_21 8000",
+				"MergeJoin_27 StreamAgg_9 TableReader_14,IndexReader_22 root left outer semi join, equal:[eq(test.t1.c1, test.t2.c1)], asc, left key:test.t1.c1, right key:test.t2.c1 8000",
```
Comment (winoros) [inline plan/explain_test.go]: Distinguish `MergeSemiJoin` from `MergeJoin`

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -58,15 +58,15 @@ func evalAstExpr(expr ast.ExprNode, ctx context.Context) (types.Datum, error) {
...
+func (b *logicalPlanBuilder) rewrite(expr ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool) (
 	expression.Expression, LogicalPlan, error) {
 	return b.rewriteWithPreprocess(expr, p, aggMapper, asScalar, nil)
 }
 
 // rewriteWithPreprocess is for handling the situation that we need to adjust the input ast tree
 // before really using its node in `expressionRewriter.Leave`. In that case, we first call
 // er.preprocess(expr), which returns a new expr. Then we use the new expr in `Leave`.
-func (b *planBuilder) rewriteWithPreprocess(expr ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool, preprocess func(ast.Node) ast.Node) (
+func (b *logicalPlanBuilder) rewriteWithPreprocess(expr ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool, preprocess func(ast.Node) ast.Node) (
```
Comment (winoros) [inline plan/expression_rewriter.go]: Do the rename in a single pr.

**Comment by zz-jason** [comment]
@coocood @hanfei1991 @winoros PTAL

File: executor/merge_join.go [hunk 3]
```diff
@@ -410,10 +237,13 @@ func (e *MergeJoinExec) computeJoin() (bool, error) {
 		var compareResult int
 		var err error
 		if e.leftRows == nil || e.rightRows == nil {
-			if e.leftRows != nil && e.rightRows == nil && e.preserveLeft {
-				// left remains and left outer join
-				// -1 will make loop continue for left
-				compareResult = -1
+			if e.leftRows != nil && e.rightRows == nil {
```
Comment (hanfei1991) [inline executor/merge_join.go]: if left is not nil, right must be nil.

File: executor/merge_join.go [hunk 4]
```diff
@@ -410,10 +237,13 @@ func (e *MergeJoinExec) computeJoin() (bool, error) {
...
+				switch e.joinType {
+				case plan.LeftOuterJoin, plan.RightOuterJoin, plan.SemiJoin, plan.LeftOuterSemiJoin:
```
Comment (hanfei1991) [inline executor/merge_join.go]: I think SemiJoin needn't to be considered ?

File: plan/physical_plans.go [hunk 5]
```diff
@@ -528,6 +528,9 @@ type PhysicalMergeJoin struct {
 
 	leftKeys  []*expression.Column
 	rightKeys []*expression.Column
+
+	// for semi joins
+	Anti bool
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply plan/physical_plans.go]: `Anti` is used by `SemiJoin` and `LeftOuterSemiJoin`, we also need to add `AntiLeftOuterSemiJoin`

[the remark this answers is not in the document]
Comment (hanfei1991) [reply plan/physical_plans.go]: I think it's no problem

File: executor/join_builder.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/plan"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type joinBuilder struct {
```
Comment (shenli) [inline executor/join_builder.go]: Could you add a few comment in this file?

File: executor/join_result_outputers.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+var (
+	_ joinResultOutputer = &mergeJoinOutputer4SemiJoin{}
+	_ joinResultOutputer = &mergeJoinOutputer4AntiSemiJoin{}
+	_ joinResultOutputer = &mergeJoinOutputer4LeftOuterSemiJoin{}
+	_ joinResultOutputer = &mergeJoinOutputer4LeftOuterJoin{}
+	_ joinResultOutputer = &mergeJoinOutputer4RightOuterJoin{}
+	_ joinResultOutputer = &mergeJoinOutputer4InnerJoin{}
+)
+
+type joinResultOutputer interface {
```
Comment (shenli) [inline executor/join_result_outputers.go]: Add comment for this interface.

**Comment by hanfei1991** [comment]
LGTM

File: plan/explain_test.go (hunk 1 above)
Comment (zz-jason) [reply plan/explain_test.go]: I distinguished them by operator name in **plan/stringer.go**. In explain info we already have join type to distinguish them, so I keep the operator name to "MergeJoin"

**Comment by zz-jason** [comment]
@winoros PTAL

File: executor/join_result_outputers.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+	_ joinResultGenerator = &mergeJoinResultGenerator4AntiSemiJoin{}
+	_ joinResultGenerator = &mergeJoinResultGenerator4LeftOuterSemiJoin{}
+	_ joinResultGenerator = &mergeJoinResultGenerator4AntiLeftOuterSemiJoin{}
+	_ joinResultGenerator = &mergeJoinResultGenerator4LeftOuterJoin{}
+	_ joinResultGenerator = &mergeJoinResultGenerator4RightOuterJoin{}
+	_ joinResultGenerator = &mergeJoinResultGenerator4InnerJoin{}
+)
+
+type joinResultGenerator interface {
+	emitMatchedInners(outer Row, inner []Row, resultBuffer []Row) ([]Row, error)
```
Comment (winoros) [inline executor/join_result_outputers.go]: add comment for each function.

File: executor/merge_join.go [hunk 9]
```diff
@@ -432,64 +236,65 @@ func (e *MergeJoinExec) computeJoin() (bool, error) {
...
 				return false, errors.Trace(err)
 			}
-		} else if compareResult < 0 {
-			initLen := len(e.outputBuf)
-			err := e.tryOutputLeftRows()
-			if err != nil {
-				return false, errors.Trace(err)
-			}
-			e.leftRows, err = e.leftRowBlock.nextBlock()
+			continue
```
Comment (winoros) [inline executor/merge_join.go]: why not use if-else?

File: executor/merge_join.go [hunk 10]
```diff
@@ -432,64 +236,65 @@ func (e *MergeJoinExec) computeJoin() (bool, error) {
...
-			initLen := len(e.outputBuf)
+			continue
+		}
 
-			// Compute cross product when both sides matches
-			err := e.computeCrossProduct()
-			if err != nil {
-				return false, errors.Trace(err)
-			}
+		// key matched, try join with other conditions
```
Comment (winoros) [inline executor/merge_join.go]: key -> Key, try -> try to

**Comment by hanfei1991** [comment]
`executor/join_result_outputers.go` , the file name also need to change.

**Comment by zz-jason** [comment]
@coocood PTAL

File: executor/join_result_generators.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+	case plan.InnerJoin:
+		return &mergeJoinResultGenerator4InnerJoin{
+			ctx:    ctx,
+			filter: filter,
+		}
+	}
+	panic("unsupported join type in func newMergeJoinResultGenerator()")
+}
+
+type mergeJoinResultGenerator4SemiJoin struct{}
```
Comment (coocood) [inline executor/join_result_generators.go]: How about `semiJoinResultGenerator`?

**Comment by winoros** [comment]
What about change `joinType` to bit representation **in the future**? We can use a bit to simply represent whether it's semi join and another bit to represent which side is its outer table. Since we have index semi join, index anti semi join and so one in the future.
LGTM

**Comment by coocood** [comment]
LGTM

Nice refactor!

**Comment by zz-jason** [comment]
this pr should be merged together with: [#375](https://github.com/pingcap/tidb-test/pull/375)

**Pull request merged**